# Lab1-Assignment

Copyright: Vrije Universiteit Amsterdam, Faculty of Humanities, CLTL

This notebook describes the assignment for Lab 1 of the text mining course. 

**Points**: each exercise is prefixed with the number of points you can obtain for the exercise.

We assume you have worked through the following notebooks:
* **Lab1.1-introduction**
* **Lab1.2-introduction-to-NLTK**
* **Lab1.3-introduction-to-spaCy** 

In this assignment, you will process an English text (**Lab1-apple-samsung-example.txt**) with both NLTK and spaCy and discuss the similarities and differences.

## Credits
The notebooks in this block have been originally created by [Marten Postma](https://martenpostma.github.io). Adaptations were made by [Filip Ilievski](http://ilievski.nl).

## Tip: how to read a file from disk
Let's open the file **Lab1-apple-samsung-example.txt** from disk.

In [2]:
from pathlib import Path

In [3]:
cur_dir = Path().resolve() # this should provide you with the folder in which this notebook is placed
path_to_file = Path.joinpath(cur_dir, 'Lab1-apple-samsung-example.txt')
print(path_to_file)
print('does path exist? ->', Path.exists(path_to_file))

/Users/timmytimtim/Documents/College/Junior 2/Text Mining/text-mining-SP26/Labs/lab1/Lab1-apple-samsung-example.txt
does path exist? -> True


If the output from the code cell above states that **does path exist? -> False**, please check that the file **Lab1-apple-samsung-example.txt** is in the same directory as this notebook.

In [4]:
with open(path_to_file) as infile:
    text = infile.read()

print('number of characters', len(text))

number of characters 1139


## [total points: 4] Exercise 1: NLTK
In this exercise, we use NLTK to apply **Part-of-speech (POS) tagging**, **Named Entity Recognition (NER)**, and **Constituency parsing**. The following code snippet already performs sentence splitting and tokenization. 

In [78]:
import nltk
from nltk.tokenize import sent_tokenize
from nltk import word_tokenize
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to
[nltk_data]    |     /Users/timmytimtim/nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to
[nltk_data]    |     /Users/timmytimtim/nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /Users/timmytimtim/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger is already up-
[nltk_data]    |       to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     /Users/timmytimtim/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_eng is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /Users/timmytimtim/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_ru is a

True

In [79]:
sentences_nltk = sent_tokenize(text)

In [80]:
tokens_per_sentence = []
for sentence_nltk in sentences_nltk:
    sent_tokens = word_tokenize(sentence_nltk)
    tokens_per_sentence.append(sent_tokens)

We will use lists to keep track of the output of the NLP tasks. We can hence inspect the output for each task using the index of the sentence.

In [81]:
sent_id = 1
print('SENTENCE', sentences_nltk[sent_id])
print('TOKENS', tokens_per_sentence[sent_id])

SENTENCE The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean system, the Galaxy Tab 8.9 Wifi tablet, the Galaxy Tab 2 10.1, Galaxy Rugby Pro and Galaxy S III mini.
TOKENS ['The', 'six', 'phones', 'and', 'tablets', 'affected', 'are', 'the', 'Galaxy', 'S', 'III', ',', 'running', 'the', 'new', 'Jelly', 'Bean', 'system', ',', 'the', 'Galaxy', 'Tab', '8.9', 'Wifi', 'tablet', ',', 'the', 'Galaxy', 'Tab', '2', '10.1', ',', 'Galaxy', 'Rugby', 'Pro', 'and', 'Galaxy', 'S', 'III', 'mini', '.']


### [point: 1] Exercise 1a: Part-of-speech (POS) tagging
Use `nltk.pos_tag` to perform part-of-speech tagging on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [82]:
pos_tags_per_sentence = []
for tokens in tokens_per_sentence:
    curr_postag = nltk.pos_tag(tokens)
    pos_tags_per_sentence.append(curr_postag)
    print(curr_postag)

[('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), ('San', 'NNP'), ('Jose', 'NNP'), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), ('California', 'NNP'), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')]
[('The', 'DT'), ('six', 'CD'), ('phones', 'NNS'), ('and', 'CC'), ('tablets', 'NNS'), ('affected', 'VBN'), ('are', 'VBP'), ('the', 'DT'), ('Galaxy', 'NNP'), ('S', 'NNP'), ('III', 'NNP'

In [83]:
print(pos_tags_per_sentence)

[[('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), ('San', 'NNP'), ('Jose', 'NNP'), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), ('California', 'NNP'), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')], [('The', 'DT'), ('six', 'CD'), ('phones', 'NNS'), ('and', 'CC'), ('tablets', 'NNS'), ('affected', 'VBN'), ('are', 'VBP'), ('the', 'DT'), ('Galaxy', 'NNP'), ('S', 'NNP'), ('III', 'NN

### [point: 1] Exercise 1b: Named Entity Recognition (NER)
Use `nltk.chunk.ne_chunk` to perform Named Entity Recognition (NER) on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [84]:
from nltk.chunk import ne_chunk
ner_tags_per_sentence = []

for tokens_pos_tagged in pos_tags_per_sentence:
    tokens_pos_tagged_and_named_entities = ne_chunk(tokens_pos_tagged)
    ner_tags_per_sentence.append(tokens_pos_tagged_and_named_entities)
    print("")
    print('ORIGINAL SENTENCE', tokens_pos_tagged)
    print('NAMED ENTITY RECOGNITION OUTPUT', tokens_pos_tagged_and_named_entities)


ORIGINAL SENTENCE [('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), ('San', 'NNP'), ('Jose', 'NNP'), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), ('California', 'NNP'), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')]
NAMED ENTITY RECOGNITION OUTPUT (S
  https/NN
  :/:
  //www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html/JJ


In [85]:
print(ner_tags_per_sentence)

[Tree('S', [('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), Tree('ORGANIZATION', [('San', 'NNP'), ('Jose', 'NNP')]), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), Tree('GPE', [('California', 'NNP')]), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), Tree('ORGANIZATION', [('Samsung', 'NNP')]), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), Tree('GPE', [('Bean', 'NNP')]), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), Tree('PERSON', [('Apple', 'NNP')]), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')]), Tree('S', [('The', 'DT'), ('six', 'CD'), ('phones', 'NNS'), ('and', 'CC'),

### [points: 2] Exercise 1c: Constituency parsing
Use the `nltk.RegexpParser` to perform constituency parsing on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [86]:
constituent_parser = nltk.RegexpParser('''
NP: {<DT>? <JJ>* <NN>*} # NP
P: {<IN>}           # Preposition
V: {<V.*>}          # Verb
PP: {<P> <NP>}      # PP -> P NP
VP: {<V> <NP|PP>*}  # VP -> V (NP|PP)*''')

In [87]:
constituency_output_per_sentence = []

for tokens_pos_tagged in pos_tags_per_sentence:
    constituent_structure = constituent_parser.parse(tokens_pos_tagged)
    constituency_output_per_sentence.append(constituent_structure)

for tree in constituency_output_per_sentence:
    print(tree) # print the sentence structure

(S
  (NP https/NN)
  :/:
  (NP
    //www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html/JJ)
  Documents/NNS
  (VP (V filed/VBN))
  to/TO
  (NP the/DT)
  San/NNP
  Jose/NNP
  (NP federal/JJ court/NN)
  (P in/IN)
  California/NNP
  (P on/IN)
  November/NNP
  23/CD
  (NP list/NN)
  six/CD
  Samsung/NNP
  products/NNS
  (VP (V running/VBG) (NP the/DT))
  ``/``
  Jelly/RB
  Bean/NNP
  ''/''
  and/CC
  ``/``
  Ice/NNP
  Cream/NNP
  Sandwich/NNP
  ''/''
  (VP (V operating/VBG))
  systems/NNS
  ,/,
  which/WDT
  Apple/NNP
  (VP (V claims/VBZ))
  (VP (V infringe/VB))
  its/PRP$
  patents/NNS
  ./.)
(S
  (NP The/DT)
  six/CD
  phones/NNS
  and/CC
  tablets/NNS
  (VP (V affected/VBN))
  (VP (V are/VBP) (NP the/DT))
  Galaxy/NNP
  S/NNP
  III/NNP
  ,/,
  (VP (V running/VBG) (NP the/DT new/JJ))
  Jelly/NNP
  Bean/NNP
  (NP system/NN)
  ,/,
  (NP the/DT)
  Galaxy/NNP
  Tab/NNP
  8.9/CD
  Wifi/NNP
  (NP tablet/NN)
  ,/,
  (NP the/DT)
  Galaxy/NNP


In [88]:
print(constituency_output_per_sentence)

[Tree('S', [Tree('NP', [('https', 'NN')]), (':', ':'), Tree('NP', [('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ')]), ('Documents', 'NNS'), Tree('VP', [Tree('V', [('filed', 'VBN')])]), ('to', 'TO'), Tree('NP', [('the', 'DT')]), ('San', 'NNP'), ('Jose', 'NNP'), Tree('NP', [('federal', 'JJ'), ('court', 'NN')]), Tree('P', [('in', 'IN')]), ('California', 'NNP'), Tree('P', [('on', 'IN')]), ('November', 'NNP'), ('23', 'CD'), Tree('NP', [('list', 'NN')]), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), Tree('VP', [Tree('V', [('running', 'VBG')]), Tree('NP', [('the', 'DT')])]), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), Tree('VP', [Tree('V', [('operating', 'VBG')])]), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), Tree('VP', [Tree('V', [('claims', 'VBZ')])]), Tree('VP', [Tree('V', [

Augment the RegexpParser so that it also detects Named Entity Phrases (NEP), e.g., that it detects *Galaxy S III* and *Ice Cream Sandwich*

In [73]:
constituent_parser_v2 = nltk.RegexpParser('''
NEP: {<NNP>+}       # Named Entity Phrase -> one or more proper nouns
NP: {<DT>? <JJ>* <NN>*} # NP
P: {<IN>}           # Preposition
V: {<V.*>}          # Verb
PP: {<P> <NP>}      # PP -> P NP
VP: {<V> <NP|PP>*}  # VP -> V (NP|PP)*''')

In [74]:
constituency_v2_output_per_sentence = []
for tokens_pos_tagged in pos_tags_per_sentence:
    constituent_structure = constituent_parser_v2.parse(tokens_pos_tagged)
    constituency_v2_output_per_sentence.append(constituent_structure)

for tree in constituency_v2_output_per_sentence:
    print(tree) # print the sentence structure

(S
  (NP https/NN)
  :/:
  (NP
    //www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html/JJ)
  Documents/NNS
  (VP (V filed/VBN))
  to/TO
  (NP the/DT)
  (NEP San/NNP Jose/NNP)
  (NP federal/JJ court/NN)
  (P in/IN)
  (NEP California/NNP)
  (P on/IN)
  (NEP November/NNP)
  23/CD
  (NP list/NN)
  six/CD
  (NEP Samsung/NNP)
  products/NNS
  (VP (V running/VBG) (NP the/DT))
  ``/``
  Jelly/RB
  (NEP Bean/NNP)
  ''/''
  and/CC
  ``/``
  (NEP Ice/NNP Cream/NNP Sandwich/NNP)
  ''/''
  (VP (V operating/VBG))
  systems/NNS
  ,/,
  which/WDT
  (NEP Apple/NNP)
  (VP (V claims/VBZ))
  (VP (V infringe/VB))
  its/PRP$
  patents/NNS
  ./.)
(S
  (NP The/DT)
  six/CD
  phones/NNS
  and/CC
  tablets/NNS
  (VP (V affected/VBN))
  (VP (V are/VBP) (NP the/DT))
  (NEP Galaxy/NNP S/NNP III/NNP)
  ,/,
  (VP (V running/VBG) (NP the/DT new/JJ))
  (NEP Jelly/NNP Bean/NNP)
  (NP system/NN)
  ,/,
  (NP the/DT)
  (NEP Galaxy/NNP Tab/NNP)
  8.9/CD
  (NEP Wifi/NNP

In [71]:
print(constituency_v2_output_per_sentence)

[Tree('S', [Tree('NP', [('https', 'NN')]), (':', ':'), Tree('NP', [('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ')]), ('Documents', 'NNS'), Tree('VP', [Tree('V', [('filed', 'VBN')])]), ('to', 'TO'), Tree('NP', [('the', 'DT')]), Tree('NEP', [('San', 'NNP'), ('Jose', 'NNP')]), Tree('NP', [('federal', 'JJ'), ('court', 'NN')]), Tree('P', [('in', 'IN')]), Tree('NEP', [('California', 'NNP')]), Tree('P', [('on', 'IN')]), Tree('NEP', [('November', 'NNP')]), ('23', 'CD'), Tree('NP', [('list', 'NN')]), ('six', 'CD'), Tree('NEP', [('Samsung', 'NNP')]), ('products', 'NNS'), Tree('VP', [Tree('V', [('running', 'VBG')]), Tree('NP', [('the', 'DT')])]), ('``', '``'), ('Jelly', 'RB'), Tree('NEP', [('Bean', 'NNP')]), ("''", "''"), ('and', 'CC'), ('``', '``'), Tree('NEP', [('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP')]), ("''", "''"), Tree('VP', [Tree('V', [('operating', 'VBG')])]), ('systems', 'NNS'), (',', ','), ('which', 'WDT'

## [total points: 1] Exercise 2: spaCy
Use Spacy to process the same text as you analyzed with NLTK.

In [69]:
import spacy
try:
    nlp = spacy.load('en_core_web_sm')
except OSError:
    !python -m spacy download en_core_web_sm
    nlp = spacy.load('en_core_web_sm')

In [70]:
doc = nlp(text)
sents = list(doc.sents)

small tip: You can use **sents = list(doc.sents)** to be able to use the index to access a sentence like **sents[2]** for the third sentence.


In [68]:
# Sentence splitting & tokenization
for sent in sents:
    print("")
    print(sent)
    for token in sent:
        print(token.text)


https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html

Documents filed to the San Jose federal court in California on November 23 list six Samsung products running the "Jelly Bean" and "Ice Cream Sandwich" operating systems, which Apple claims infringe its patents.

https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html



Documents
filed
to
the
San
Jose
federal
court
in
California
on
November
23
list
six
Samsung
products
running
the
"
Jelly
Bean
"
and
"
Ice
Cream
Sandwich
"
operating
systems
,
which
Apple
claims
infringe
its
patents
.



The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean system, the Galaxy Tab 8.9 Wifi tablet, the Galaxy Tab 2 10.1, Galaxy Rugby Pro and Galaxy S III mini.

The
six
phones
and
tablets
affected
are
the
Galaxy
S
III
,
running
the
new
Jelly
Bean
system
,
the
Galaxy
Tab
8.9
Wifi
tablet
,
the
Galaxy
Tab
2
10.

In [67]:
# Part of speech tagging
for sent in sents:
    print("")
    for token in sent:
        print(token.text, token.pos_, token.tag_)


https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html PROPN NNP


 SPACE _SP
Documents PROPN NNPS
filed VERB VBD
to ADP IN
the DET DT
San PROPN NNP
Jose PROPN NNP
federal ADJ JJ
court NOUN NN
in ADP IN
California PROPN NNP
on ADP IN
November PROPN NNP
23 NUM CD
list NOUN NN
six NUM CD
Samsung PROPN NNP
products NOUN NNS
running VERB VBG
the DET DT
" PUNCT ``
Jelly PROPN NNP
Bean PROPN NNP
" PUNCT ''
and CCONJ CC
" PUNCT ``
Ice PROPN NNP
Cream PROPN NNP
Sandwich PROPN NNP
" PUNCT ''
operating NOUN NN
systems NOUN NNS
, PUNCT ,
which PRON WDT
Apple PROPN NNP
claims VERB VBZ
infringe VERB VBP
its PRON PRP$
patents NOUN NNS
. PUNCT .

 SPACE _SP

The DET DT
six NUM CD
phones NOUN NNS
and CCONJ CC
tablets NOUN NNS
affected VERB VBN
are AUX VBP
the DET DT
Galaxy PROPN NNP
S PROPN NNP
III PROPN NNP
, PUNCT ,
running VERB VBG
the DET DT
new ADJ JJ
Jelly PROPN NNP
Bean PROPN NNP
system NOUN NN
, PUNCT ,
the DET DT
Galaxy PROPN NNP
Tab 

In [66]:
# Named Entity Recognition (NER)
for ent in doc.ents:
    print(ent.text, ent.label_)

San Jose GPE
California GPE
November 23 DATE
six CARDINAL
Samsung ORG
Jelly Bean WORK_OF_ART
Apple ORG
six CARDINAL
the Galaxy S III PERSON
Jelly Bean ORG
Galaxy Tab 8.9 Wifi PERSON
Galaxy Tab 2 10.1 PERSON
Apple ORG
Apple ORG
August DATE
Samsung ORG
US GPE
Apple ORG
1.05bn MONEY
0.66bn MONEY
iPad ORG
iPhone ORG
Samsung ORG
UK GPE
Samsung ORG
Apple ORG
South Korean NORP
iPad ORG


In [65]:
# NER visualization (from Lab1.3)
from spacy import displacy
displacy.render(doc, jupyter=True, style='ent')

In [59]:
# Dependency parsing
for sent in sents:
    print("")
    for token in sent:
        print(token.text, token.dep_, token.head)


https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html compound Documents


 dep https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html
Documents nsubj filed
filed ROOT filed
to prep filed
the det court
San compound Jose
Jose nmod court
federal amod court
court pobj to
in prep court
California pobj in
on prep filed
November pobj on
23 nummod November
list npadvmod filed
six nummod products
Samsung compound products
products appos list
running acl products
the det Bean
" punct Bean
Jelly compound Bean
Bean dobj running
" punct Bean
and cc Bean
" punct Sandwich
Ice compound Cream
Cream compound Sandwich
Sandwich nmod systems
" punct Sandwich
operating compound systems
systems conj Bean
, punct systems
which nsubj claims
Apple nsubj claims
claims relcl systems
infringe xcomp claims
its poss patents
patents dobj infringe
. punct filed

 dep .

The det phones
six nummod phone

## [total points: 7] Exercise 3: Comparison NLTK and spaCy
We will now compare the output of NLTK and spaCy, i.e., in what do they differ?

### [points: 3] Exercise 3a: Part of speech tagging
Compare the output from NLTK and spaCy regarding part of speech tagging.

* To compare, you probably would like to compare sentence per sentence. Describe if the sentence splitting is different for NLTK than for spaCy. If not, where do they differ?
* After checking the sentence splitting, select a sentence for which you expect interesting results and perhaps differences. Motivate your choice.
* Compare the output in `token.tag` from spaCy to the part of speech tagging from NLTK for each token in your selected sentence. Are there any differences? This is not a trick question; it is possible that there are no differences.

In [64]:
# Sentence splitting comparison
print(f"Number of NLTK sentences: {len(sentences_nltk)}")
print(f"Number of spaCy sentences: {len(sents)}")

print("\nNLTK sentences:")
for i, s in enumerate(sentences_nltk):
    print(f"  [{i}]: {s}")

print("\nspaCy sentences:")
for i, s in enumerate(sents):
    clean = s.text.strip()
    print(f"  [{i}]: {clean}")

# POS comparison for "The six phones and tablets affected..." sentence
target = "The six phones"
nltk_idx = next(i for i, s in enumerate(sentences_nltk) if target in s)
spacy_idx = next(i for i, s in enumerate(sents) if target in s.text)

print(f"\n=== POS Tag Comparison for sentence (NLTK idx={nltk_idx}, spaCy idx={spacy_idx}) ===\n")

spacy_tokens = [(t.text, t.tag_) for t in sents[spacy_idx] if t.text.strip()]
nltk_tokens = pos_tags_per_sentence[nltk_idx]

print(f"{'Token':<15} {'NLTK':<10} {'spaCy':<10} {'Match'}")
print("-" * 45)
differences = 0
for i in range(min(len(nltk_tokens), len(spacy_tokens))):
    nt, ntag = nltk_tokens[i]
    st, stag = spacy_tokens[i]
    match = "Yes" if ntag == stag else "NO  <--"
    if ntag != stag:
        differences += 1
    print(f"{nt:<15} {ntag:<10} {stag:<10} {match}")

print(f"\nTotal differences: {differences}")

Number of NLTK sentences: 6
Number of spaCy sentences: 6

NLTK sentences:
  [0]: https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html

Documents filed to the San Jose federal court in California on November 23 list six Samsung products running the "Jelly Bean" and "Ice Cream Sandwich" operating systems, which Apple claims infringe its patents.
  [1]: The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean system, the Galaxy Tab 8.9 Wifi tablet, the Galaxy Tab 2 10.1, Galaxy Rugby Pro and Galaxy S III mini.
  [2]: Apple stated it had “acted quickly and diligently" in order to "determine that these newly released products do infringe many of the same claims already asserted by Apple."
  [3]: In August, Samsung lost a US patent case to Apple and was ordered to pay its rival $1.05bn (£0.66bn) in damages for copying features of the iPad and iPhone in its Galaxy range of devices.
  [4]: Samsung, which is

**Sentence splitting:**

Both NLTK and spaCy produce exactly **6 sentences**, and the sentence boundaries are nearly identical. The only difference is in **quote handling** around sentence 2/3. Sentence 2 ends with `Apple."`, the period is inside the closing quotation mark. NLTK recognizes `."` as a sentence ending and attaches the closing quote to sentence 2, so sentence 3 cleanly starts with `In August...`. spaCy, however, splits the period and quotation mark into separate tokens and does not attach the closing `"` to sentence 2. Instead, that orphaned `"` gets pushed to the start of sentence 3, resulting in `" In August, Samsung lost...`. This shows that spaCy's sentence splitter struggles with punctuation sequences like `."` at sentence boundaries.

Both tools merge the URL with the "Documents filed..." paragraph into a single sentence (sentence 0). This is because sentence tokenizers rely on punctuation (`.`, `?`, `!`) to detect boundaries: the URL ending in `.html` is not recognized as a sentence-ending period, and the blank line (newline characters) between the URL and the paragraph is not treated as a sentence boundary either. This is a limitation when processing text that contains non-natural-language elements like URLs.

**Sentence choice:**

We selected sentence index 1 ("The six phones and tablets affected are the Galaxy S III...") because it contains many product names (Galaxy S III, Jelly Bean, Galaxy Tab), version numbers (8.9, 10.1), and a mix of proper nouns, common nouns, and adjectives, making it a good candidate for revealing POS tagging differences.

**POS tag comparison:**

For the selected sentence, the fine-grained POS tags (`tag_`) from both NLTK and spaCy are **identical** for all 41 tokens (0 differences). Both tools use the Penn Treebank tagset and fully agree on this sentence.

However, differences do appear in other sentences. For sentence 0 ("Documents filed to the San Jose federal court..."), where the URL causes different tokenization, comparing the shared tokens reveals:
- **"Jelly"**: NLTK → RB (adverb), spaCy → NNP (proper noun). spaCy is correct here since "Jelly Bean" is a proper noun (Android OS version name).
- **"Documents"**: NLTK → NNS (common plural noun), spaCy → NNPS (proper plural noun).
- **"filed"**: NLTK → VBN (past participle), spaCy → VBD (simple past). NLTK arguably captures the reduced relative clause ("Documents [that were] filed") better.
- **"operating"**: NLTK → VBG (gerund), spaCy → NN (noun). In "operating systems," spaCy treats it as part of a compound noun.

These differences show that while both tools use the same tagset, their underlying models can disagree on ambiguous words.

### [points: 2] Exercise 3b: Named Entity Recognition (NER)
* Describe differences between the output from NLTK and spaCy for Named Entity Recognition. Which one do you think performs better?

**Differences between NLTK and spaCy NER:**

1. **Entity type richness**: spaCy uses a much richer set of entity labels: ORG, GPE, DATE, MONEY, CARDINAL, NORP (nationality/religious/political group), and WORK_OF_ART. NLTK uses a smaller set: ORGANIZATION, PERSON, GPE, and GSP.

2. **Multi-word entities**: spaCy correctly groups multi-word entities (e.g., "November 23" as DATE, "Galaxy Tab 8.9 Wifi" as one entity, "South Korean" as NORP). NLTK typically recognizes only individual words or short spans (e.g., "Samsung" alone, "Galaxy" alone).

3. **Numeric/value entities**: spaCy recognizes numbers ("six" as CARDINAL), dates ("August" as DATE), and monetary values ("$1.05bn" as MONEY, "£0.66bn" as MONEY). NLTK misses almost all of these.

4. **Misclassifications**:
   - NLTK labels "Apple" as PERSON in some sentences instead of ORGANIZATION.
   - NLTK labels "Bean" (from "Jelly Bean") as GPE (a location).
   - NLTK labels "August" as GPE.
   - spaCy mislabels product names like "the Galaxy S III" and "Galaxy Tab 8.9 Wifi" as PERSON instead of a product entity.
   - spaCy labels "Jelly Bean" as WORK_OF_ART in one instance and ORG in another: inconsistent.

**Which performs better?**

spaCy performs significantly better overall. It recognizes more entities, groups multi-word entities correctly, uses a richer label set, and makes fewer critical misclassifications. While it incorrectly labels some Samsung product names as PERSON, its overall accuracy and coverage far exceed NLTK's rule-based NER module.

### [points: 2] Exercise 3c: Constituency/dependency parsing
Choose one sentence from the text and run constituency parsing using NLTK and dependency parsing using spaCy.
* describe briefly the difference between constituency parsing and dependency parsing
* describe differences between the output from NLTK and spaCy.

In [63]:
# Choose "Samsung, which is the world's top mobile phone maker, is appealing the ruling."
target = "world"
nltk_idx = next(i for i, s in enumerate(sentences_nltk) if target in s)
spacy_idx = next(i for i, s in enumerate(sents) if target in s.text)

print("Selected sentence:", sentences_nltk[nltk_idx].strip())

print("\n=== NLTK Constituency Parsing (v2, with NEP rule) ===\n")
print(constituency_v2_output_per_sentence[nltk_idx])

print("\n=== spaCy Dependency Parsing ===\n")
print(f"{'Token':<15} {'Dep Label':<12} {'Explanation':<30} {'Head'}")
print("-" * 70)
for token in sents[spacy_idx]:
    if token.text.strip():
        explanation = spacy.explain(token.dep_) or ""
        print(f"{token.text:<15} {token.dep_:<12} {explanation:<30} {token.head.text}")

print("\n=== spaCy Dependency Tree Visualization ===")
displacy.render(sents[spacy_idx], jupyter=True, style='dep')

Selected sentence: Samsung, which is the world's top mobile phone maker, is appealing the ruling.

=== NLTK Constituency Parsing (v2, with NEP rule) ===

(S
  (NEP Samsung/NNP)
  ,/,
  which/WDT
  (VP (V is/VBZ) (NP the/DT world/NN))
  's/POS
  (NP top/JJ mobile/NN phone/NN maker/NN)
  ,/,
  (VP (V is/VBZ))
  (VP (V appealing/VBG) (NP the/DT ruling/NN))
  ./.)

=== spaCy Dependency Parsing ===

Token           Dep Label    Explanation                    Head
----------------------------------------------------------------------
Samsung         nsubj        nominal subject                appealing
,               punct        punctuation                    Samsung
which           nsubj        nominal subject                is
is              relcl        relative clause modifier       Samsung
the             det          determiner                     world
world           poss         possession modifier            maker
's              case         case marking                   world

**Constituency parsing vs. dependency parsing:**

- **Constituency parsing** decomposes a sentence into nested sub-phrases (constituents) in a tree structure. Each internal node represents a phrase type (NP, VP, PP), and the leaves are individual words. It answers: *"How do words group into phrases?"*

- **Dependency parsing** identifies directed grammatical relationships between individual words. Each word (dependent) is connected to exactly one other word (head) via a labeled relation (e.g., nsubj, dobj, amod). It answers: *"Which word depends on which other word?"*

**Output differences for the selected sentence:**

NLTK's `RegexpParser` (v2, with our added NEP rule `{<NNP>+}`) produces a rule-based constituency tree. It captures phrase structures that match the regex patterns (NP, VP, PP, NEP), and the NEP rule successfully groups "Samsung" as a named entity phrase. However, many tokens are still left unchunked directly under the root `S` node — function words like "which" and the possessive "'s" are not grouped into any phrase. The output remains incomplete and shallow.

spaCy's dependency parser produces a fully connected graph where *every* token has an assigned head and dependency label. Using `spacy.explain()` (as taught in Lab1.3), we can interpret each label. It correctly identifies:
- "Samsung" as the subject (nsubj) of "appealing"
- "which" as subject (nsubj) of "is" in the relative clause
- "maker" as the attribute (attr) of the copular "is"
- "ruling" as the direct object (dobj) of "appealing"
- Possessive structure: "'s" marks "world" as the possessor of "maker"

The `displacy.render()` visualization (also from Lab1.3) makes the dependency structure much easier to read than the raw text output, clearly showing the directed arcs between tokens.

The spaCy output is more informative and accurate because it uses a trained statistical model that handles complex constructions (relative clauses, possessives, appositives), while NLTK's rule-based approach is limited to what the handwritten regex rules can capture.

# End of this notebook